pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py

```diff
@@ -92,72 +91,67 @@ def __init__(self, fx_name):
+        return len(_inter)
 
     def check_dataloader_idx(self, result: Result) -> bool:
-        add_dataloader_idx = False
-        try:
-            if len(result.keys()) > 1:
-                random_key = [*result.keys()][-1]
-                add_dataloader_idx = result["meta"][random_key]["dataloader_idx"] is not None
-                return add_dataloader_idx
-            return add_dataloader_idx
-        except Exception:
-            return add_dataloader_idx
-
-    def get_lastest_from_func_name(self, func_name, *args, latest=True, **kwargs):
+        random_key = [*result.keys()][-1]
+        add_dataloader_idx = result["meta"][random_key]["dataloader_idx"] is not None
+        return add_dataloader_idx
+
+    def get_lastest_from_func_name(self, latest_result, func_name: str, *args, **kwargs) -> Dict:
         results = {}
-        if latest:
-            for dl_idx in range(self.num_dataloaders):
-                dl_idx = str(dl_idx)
-                if self._internal_type == ResultStoreType.OUTSIDE_BATCH_TRAIN_LOOP:
-                    latest_result = self._internals[dl_idx][-1]
-                else:
-                    latest_result = self.get_latest_from_dict(dl_idx)
-                add_dataloader_idx = self.check_dataloader_idx(latest_result)
-                func = getattr(latest_result, func_name)
-                results.update(func(*args, add_dataloader_idx=add_dataloader_idx, **kwargs))
-            return results
-        raise NotImplementedError
+        add_dataloader_idx = self.check_dataloader_idx(latest_result)
+        func = getattr(latest_result, func_name)
+        results.update(func(*args, add_dataloader_idx=add_dataloader_idx, **kwargs))
+        return results
+
+    def run_lastest_from_func_name(self, func_name, *args, **kwargs) -> Dict:
+        """
+        This function used cache_ref and cache_result to optimize loading metrics
 
-    def get_batch_pbar_metrics(self, latest=True, *args, **kwargs):
-        return self.get_lastest_from_func_name("get_batch_pbar_metrics", *args, latest=latest, **kwargs)
+        Context: As we update the logger_connector metrics on every `self.log` call,
+        and it can be pretty time consuming, especially when logging outside batch loop.
 
-    def get_batch_log_metrics(self, latest=True, *args, **kwargs):
-        return self.get_lastest_from_func_name("get_batch_log_metrics", *args, latest=latest, **kwargs)
+        HookResultStore keeps track of its latest added result object,
+        and cache its pbar and log metrics if already called on,
+        """
+        results = []
+        for dl_idx in range(self.num_dataloaders):
+            dl_idx = str(dl_idx)
+            latest_result = self._latest_ref[dl_idx]
+            result = self.get_lastest_from_func_name(latest_result, func_name, *args, **kwargs)
+            results.append(result)
+        return dict(ChainMap(*results))
```

Comment:
Could we get an explanation of whats happening here @tchaton?

Reply:
Yes, I am getting the latest result object and get latest metrics for each dataloader.